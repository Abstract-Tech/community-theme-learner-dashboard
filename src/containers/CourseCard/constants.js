/* eslint-disable import/prefer-default-export */
export const EXECUTIVE_EDUCATION_2U = 'executive-education-2u';
