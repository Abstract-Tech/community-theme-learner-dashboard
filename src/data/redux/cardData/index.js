/* eslint-disable import/prefer-default-export */
export { default as selectors } from './selectors';
