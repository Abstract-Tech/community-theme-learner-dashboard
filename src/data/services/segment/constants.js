import { StrictDict } from 'utils';

export const events = StrictDict({
  courseEnroll: 'courseEnroll',
  entitlementUnenroll: 'entitlementUnenroll',
  sessionChange: 'sessionChange',
  unenrollReason: 'unenrollReason',
});

export const eventNames = StrictDict({
  [events.courseEnroll]: 'edx.bi.user.program-details.enrollment',
  [events.entitlementUnenroll]: 'entitlement_unenrollment_reason.selected',
  [events.sessionChange]: ({ action }) => `course-dashboard.${action}-session`, // 'switch', 'new', 'leave'
  [events.unenrollReason]: 'unenrollment_reason.selected',
});

export const trackingCategory = 'learner-home';

export const pageViewEvent = { category: trackingCategory };
