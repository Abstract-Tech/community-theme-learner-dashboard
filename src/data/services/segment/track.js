import { categories, eventNames } from './constants';
import { trackEvent } from './utils';

export const enterCourseClicked = (courseId) => () => trackEvent(
  eventNames.enterCourseClicked,
  { category: categories.dashboard, label: courseId },
);
export const courseTitleClicked = (courseId) => () => trackEvent(
  eventNames.courseTitleClicked,
  { category: categories.dashboard, label: courseId },
);
export const courseOptionsDropdownClicked = (courseId) => () => trackEvent(
  eventNames.courseOptionsDropdownClicked,
  { category: categories.dashboard, label: courseId },
);
export const courseImageClicked = (courseId) => () => trackEvent(
  eventNames.courseImageClicked,
  { category: categories.dashboard, label: courseId },
);
export const upgradeClicked = (courseId) => () => {
  trackEvent(
    eventNames.upgradeButtonClicked,
    { category: categories.upgrade, label: courseId },
  );
  trackEvent(eventNames.upgradeButtonClickedEnrollment, { location: 'learner_dashboard' });
  trackEvent(eventNames.upgradeButtonClickedUpsell, {
    linkName: 'course_dashboard_green',
    linkType: 'button',
    pageName: 'course_dashboard',
    linkCategory: 'green_update',
  });
};

export const shareClicked = (courseId, site) => () => trackEvent(
  eventNames.shareClicked,
  { course_id: courseId, social_media_site: site, location: 'dashboard' },
);

export const newEntitlementEnrollment = (selection) => () => {
  trackEvent(eventNames.newSession, { fromCourseRun: null, toCourseRun: selection });
};

export const switchEntitlementSession = (fromCourseRun, toCourseRun) => () => {
  trackEvent(eventNames.switchSession, {
    fromCourseRun,
    toCourseRun,
  });
};

export const leaveEntitlementSession = (fromCourseRun) => () => {
  trackEvent(eventNames.leaveSession, {
    fromCourseRun,
    toCourseRun: null,
  });
};

export const unenrollReason = (courseId, reason) => () => trackEvent(
  eventNames.unenrollReason,
  {
    category: categories.userEngagement,
    reason,
    displayName: 'v1',
    course_id: courseId,
  },
);

export const entitlementUnenrollReason = (courseId, reason) => () => trackEvent(
  eventNames.entitlementUnenrollReason,
  {
    category: categories.userEngagement,
    reason,
    displayName: 'v1',
  },
);

export const enterpriseDashboardModalOpened = (enterpriseUUID) => () => trackEvent(
  eventNames.enterpriseDashboardModalOpened,
  { enterpriseUUID },
);
export const enterpriseDashboardModalCTAClicked = (enterpriseUUID) => () => trackEvent(
  eventNames.enterpriseDashboardModalCTAClicked,
  { enterpriseUUID },
);
export const enterpriseDashboardModalClosed = (enterpriseUUID) => () => trackEvent(
  eventNames.enterpriseDashboardModalClosed,
  { enterpriseUUID, source: 'Escape' },
);
