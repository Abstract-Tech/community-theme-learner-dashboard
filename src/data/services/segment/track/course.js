import { categories, eventNames } from '../constants';
import { createEventTracker, createLinkTracker } from '../utils';

/** Course Link events **/
const courseEventTracker = (eventName, courseId, options = {}) => createEventTracker(
  eventName,
  { category: categories.dashboard, label: courseId, ...options },
);
/**
 * Generate a hook to allow components to provide a courseId and link href and provide
 * a link tracker with defined event name and options, over a set of default optiosn.
 * @param {string} eventName - event name for the click event
 * @param {[object]} options - (optional) event options
 * @return {callback} - component hook returning a link tracking event callback
 */
const courseLinkTracker = (eventName, options = {}) => (courseId, href) => createLinkTracker(
  courseEventTracker(eventName, options, courseId),
  href,
);

const upgradeButtonClicked = (courseId) => createEventTracker(
  eventNames.upgradeButtonClicked,
  { category: categories.upgrade, label: courseId },
);
const upgradeButtonClickedEnrollment = () => createEventTracker(
  eventNames.upgradeButtonClickedEnrollment,
  { location: 'learner_dashboard' },
);
const upgradeButtonClickedUpsell = () => createEventTracker(
  eventNames.upgradeButtonClickedUpsell,
  {
    linkName: 'course_dashboard_green',
    linkType: 'button',
    pageName: 'course_dashboard',
    linkCategory: 'green_update',
  },
);

export const courseImageClicked = courseLinkTracker(eventNames.courseImageClicked);
export const courseOptionsDropdownClicked = (courseId) => courseEventTracker(
  eventNames.courseOptionsDropdownClicked,
  courseId,
);
export const courseTitleClicked = courseLinkTracker(eventNames.courseTitleClicked);
export const enterCourseClicked = courseLinkTracker(eventNames.enterCourseClicked);
export const upgradeClicked = (courseId, href) => createLinkTracker(
  () => {
    upgradeButtonClicked(courseId);
    upgradeButtonClickedEnrollment();
    upgradeButtonClickedUpsell();
  },
  href,
);

export default {
  courseImageClicked,
  courseOptionsDropdownClicked,
  courseTitleClicked,
  enterCourseClicked,
  upgradeClicked,
};
