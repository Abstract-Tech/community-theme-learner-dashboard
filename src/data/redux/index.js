import { combineReducers } from 'redux';

import { StrictDict } from 'utils';

import * as app from './app';
import * as requests from './requests';
import * as cardData from './cardData';

export { default as thunkActions } from './thunkActions';

const modules = {
  app,
  requests,
  cardData,
};

const moduleProps = (propName) => Object.keys(modules).reduce(
  (obj, moduleKey) => {
    const value = modules[moduleKey][propName];
    return value ? { ...obj, [moduleKey]: value } : obj;
  },
  {},
);

const rootReducer = combineReducers(moduleProps('reducer'));

const actions = StrictDict(moduleProps('actions'));

const selectors = StrictDict(moduleProps('selectors'));

export { actions, selectors };

export default rootReducer;
