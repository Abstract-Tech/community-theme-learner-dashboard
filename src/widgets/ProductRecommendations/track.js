import { StrictDict } from 'utils';
import { createLinkTracker, createEventTracker } from 'data/services/segment/utils';

export const eventNames = StrictDict({
  productCardClicked: 'edx.bi.2u-product-card.clicked',
  discoveryCardClicked: 'edx.bi.user.discovery.card.click',
  recommendationsHeaderClicked: 'edx.bi.link.recommendations.header.clicked',
  recommendationsViewed: 'edx.bi.user.recommendations.viewed',
});

export const productCardClicked = (courseRunKey, courseTitle, courseType, href) => {
  createLinkTracker(
    createEventTracker(eventNames.productCardClicked, {
      catagory: 'recommender',
      courserun_key: courseRunKey,
      label: courseTitle,
      page: 'dashboard',
      product_line: courseType,
    }),
    href,
  );
};

export const discoveryCardClicked = (courseRunKey, courseTitle, href) => {
  createLinkTracker(
    createEventTracker(eventNames.discoveryCardClicked, {
      catagory: 'recommender',
      courserun_key: courseRunKey,
      label: courseTitle,
      page: 'dashboard',
      product_line: 'open-course',
    }),
    href,
  );
};

export const recommendationsHeaderClicked = (courseType, href) => {
  createLinkTracker(
    createEventTracker(eventNames.recommendationsHeaderClicked, {
      category: 'recommender',
      product_line: courseType,
      page: 'dashboard',
    }),
    href,
  );
};

export const recommendationsViewed = (isControl, courseKey) => {
  createEventTracker(eventNames.recommendationsViewed, {
    is_control: isControl,
    page: 'dashboard',
    course_key: courseKey,
  });
};
