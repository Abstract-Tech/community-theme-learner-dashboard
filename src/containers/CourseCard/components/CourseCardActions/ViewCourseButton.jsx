import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';

import track from 'tracking';
import { hooks } from 'data/redux';
import ActionButton from './ActionButton';
import messages from './messages';

export const ViewCourseButton = ({ cardId }) => {
  const { homeUrl } = hooks.useCardCourseRunData(cardId);
  const { hasAccess } = hooks.useCardEnrollmentData(cardId);
  const handleClick = hooks.useTrackCourseEvent(
    track.course.enterCourseClicked,
    cardId,
    homeUrl,
  );
  const { formatMessage } = useIntl();

  return (
    <ActionButton
      disabled={!hasAccess}
      as="a"
      href="#"
      onClick={handleClick}
    >
      {formatMessage(messages.viewCourse)}
    </ActionButton>
  );
};
ViewCourseButton.propTypes = {
  cardId: PropTypes.string.isRequired,
};
export default ViewCourseButton;
