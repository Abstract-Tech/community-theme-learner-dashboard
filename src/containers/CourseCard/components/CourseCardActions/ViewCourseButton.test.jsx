import { shallow } from 'enzyme';

import track from 'data/services/segment/track';
import { htmlProps } from 'data/constants/htmlKeys';
import { hooks } from 'data/redux';
import ViewCourseButton from './ViewCourseButton';

jest.mock('data/services/segment/track', () => ({
  course: {
    enterCourseClicked: jest.fn().mockName('segment.enterCourseClicked'),
  },
}));

jest.mock('data/redux', () => ({
  hooks: {
    useCardCourseRunData: jest.fn(),
    useCardEnrollmentData: jest.fn(),
    useCardEntitlementData: jest.fn(),
    useTrackCourseEvent: jest.fn(
      (eventName, cardId, upgradeUrl) => ({ trackCourseEvent: { eventName, cardId, upgradeUrl } }),
    ),
  },
}));
jest.mock('./ActionButton', () => 'ActionButton');

let wrapper;
const defaultProps = { cardId: 'cardId' };
const homeUrl = 'homeUrl';

const createWrapper = ({
  hasAccess = false,
  isEntitlement = false,
  isExpired = false,
  propsOveride = {},
}) => {
  hooks.useCardCourseRunData.mockReturnValue({ homeUrl });
  hooks.useCardEnrollmentData.mockReturnValueOnce({ hasAccess });
  hooks.useCardEntitlementData.mockReturnValueOnce({ isEntitlement, isExpired });
  return shallow(<ViewCourseButton {...defaultProps} {...propsOveride} />);
};

describe('ViewCourseButton', () => {
  describe('learner has access to course', () => {
    beforeEach(() => {
      wrapper = createWrapper({ hasAccess: true });
    });
    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    test('links to home URL', () => {
      expect(wrapper.prop(htmlProps.onClick)).toEqual(hooks.useTrackCourseEvent(
        track.course.enterCourseClicked,
        defaultProps.cardId,
        homeUrl,
      ));
    });
    test('link is enabled', () => {
      expect(wrapper.prop(htmlProps.disabled)).toEqual(false);
    });
  });
  describe('learner does not have access to course', () => {
    beforeEach(() => {
      wrapper = createWrapper({ hasAccess: false });
    });
    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    test('links to home URL', () => {
      expect(wrapper.prop(htmlProps.onClick)).toEqual(hooks.useTrackCourseEvent(
        track.course.enterCourseClicked,
        defaultProps.cardId,
        homeUrl,
      ));
    });
    test('link is enabled', () => {
      expect(wrapper.prop(htmlProps.disabled)).toEqual(true);
    });
  });
});
