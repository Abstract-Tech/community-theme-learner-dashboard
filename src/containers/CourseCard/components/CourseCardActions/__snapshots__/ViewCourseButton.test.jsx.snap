// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewCourseButton learner does not have access to course snapshot 1`] = `
<ActionButton
  as="a"
  disabled={true}
  href="#"
  onClick={
    Object {
      "trackCourseEvent": Object {
        "cardId": "cardId",
        "eventName": [MockFunction segment.enterCourseClicked],
        "upgradeUrl": "homeUrl",
      },
    }
  }
>
  View Course
</ActionButton>
`;

exports[`ViewCourseButton learner has access to course snapshot 1`] = `
<ActionButton
  as="a"
  disabled={false}
  href="#"
  onClick={
    Object {
      "trackCourseEvent": Object {
        "cardId": "cardId",
        "eventName": [MockFunction segment.enterCourseClicked],
        "upgradeUrl": "homeUrl",
      },
    }
  }
>
  View Course
</ActionButton>
`;
